### ISSUE TITLE
Support Spring Cloud Stream

### ISSUE THREAD

**Issue by chaudharydeepak** [spring-projects-experimental/spring-native#693]
[Application Source Code](https://github.com/chaudharydeepak/cloudstreamnative) is a Spring Cloud Stream native application:

```
<dependencies>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream-binder-kafka</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.experimental</groupId>
            <artifactId>spring-native</artifactId>
            <version>${spring-native.version}</version>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-configuration-processor</artifactId>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>org.projectlombok</groupId>
            <artifactId>lombok</artifactId>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream</artifactId>
            <scope>test</scope>
            <classifier>test-binder</classifier>
            <type>test-jar</type>
        </dependency>
    </dependencies>
```

Application runs fine on JVM, as well as the native image is generated just fine:
```
mvn clean -Pnative-image package
```

However, startup of native image throws the following exception:
```
2021-03-31 22:11:47.221 ERROR 18378 --- [           main] o.s.boot.SpringApplication               : Application run failed

com.oracle.svm.core.jdk.UnsupportedFeatureError: Proxy class defined by interfaces [interface org.springframework.cloud.stream.annotation.EnableBinding, interface org.springframework.core.annotation.SynthesizedAnnotation] not found. Generating proxy classes at runtime is not supported. Proxy classes need to be defined at image build time by specifying the list of interfaces that they implement. To define proxy classes use -H:DynamicProxyConfigurationFiles=<comma-separated-config-files> and -H:DynamicProxyConfigurationResources=<comma-separated-config-resources> options.
        at com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:87) ~[na:na]

```

Following the exception trace, I created a proxy-config.json as following:
```
[
  [
    "org.springframework.cloud.stream.annotation.EnableBinding",
    "org.springframework.core.annotation.SynthesizedAnnotation"
  ]
]
```
and provided to build:
```
...
<configuration>
                            <!-- The native image build needs to know the entry point to your application -->
                            <mainClass>com.dccorp.nativeapps.NativeCloudStreamsApplication</mainClass>
                            <buildArgs>
                                <!-- uncomment to generate dump file for graalvm dashboard analysis.-->
                                <!--                                -H:DashboardDump=dumpfileoversizedbuildArgs-->
                                <!--                                -H:+DashboardAll-->
                                <!-- you can provide additional params. -->
                                -H:DynamicProxyConfigurationFiles=classes/proxy-config.json
                                --enable-url-protocols=http
                            </buildArgs>
                        </configuration>
...
```

Now, the generated native image execution fails with following error message:
```
...
org.springframework.core.annotation.AnnotationConfigurationException: Method [public abstract java.lang.Class[] org.springframework.cloud.stream.annotation.EnableBinding.value()] is unsupported for synthesized annotation type [interface org.springframework.cloud.stream.annotation.EnableBinding]
        at org.springframework.core.annotation.SynthesizedMergedAnnotationInvocationHandler.invoke(SynthesizedMergedAnnotationInvocationHandler.java:92) ~[na:na]
...
```

Any pointers to what I might be missing here? Applications runs fine with JVM.

**Comment by sdeleuze** [comment]
Given the place where this error happen:
```
public Object invoke(Object proxy, Method method, Object[] args) {
		if (ReflectionUtils.isEqualsMethod(method)) {
			return annotationEquals(args[0]);
		}
		if (ReflectionUtils.isHashCodeMethod(method)) {
			return annotationHashCode();
		}
		if (ReflectionUtils.isToStringMethod(method)) {
			return annotationToString();
		}
		if (isAnnotationTypeMethod(method)) {
			return this.type;
		}
		if (this.attributes.indexOf(method.getName()) != -1) {
			return getAttributeValue(method);
		}
		throw new AnnotationConfigurationException(String.format(
				"Method [%s] is unsupported for synthesized annotation type [%s]", method, this.type));
	}
```
I guess you need a hint like `@TypeHint(types = org.springframework.cloud.stream.annotation.EnableBinding.class, access = AccessBits.FULL_REFLECTION)`.

**Comment by olegz** [comment]
FWIW, the EnableBinding and annotation-based programming model is deprecated and soon won't even be supported. So please switch to [functional programing model](https://docs.spring.io/spring-cloud-stream/docs/3.0.10.RELEASE/reference/html/spring-cloud-stream.html#spring-cloud-stream-overview-producing-consuming-messages) and see if you have any errors

**Comment by chaudharydeepak** [comment]
@olegz Yeah, I am using functional programming model only - there's just 1 function:
```
@Bean
    public java.util.function.Consumer<Person> appConsumer() {
        return person -> log.info("Data received {}", person);
    }
```

and spring boot / cloud versions:
```
...
<parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.4.4</version>
        <relativePath/> <!-- lookup parent from repository -->
    </parent>
...
<spring-cloud.version>2020.0.2</spring-cloud.version>
...
```

@sdeleuze - trying with `@TypeHint` results in another exception:
```
2021-04-01 08:44:41.397 ERROR 1352 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'IntegrationConfigurationBeanFactoryPostProcessor': Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.integration.config.IntegrationConfigurationBeanFactoryPostProcessor]: No default constructor found; nested exception is java.lang.NoSuchMethodException: org.springframework.integration.config.IntegrationConfigurationBeanFactoryPostProcessor.<init>()
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateBean(AbstractAutowireCapableBeanFactory.java:1316) ~[na:na]

```

**Comment by olegz** [comment]
@chaudharydeepak any chance you can post a whole stack trace? I am a bit puzzled as to why are you getting it if you are not using it (the EnableBinding that is)

**Comment by chaudharydeepak** [comment]
Sure @chaudharydeepak - attached here is the FINE log. Thank you.
[app.log](https://github.com/spring-projects-experimental/spring-native/files/6244481/app.log)

**Comment by olegz** [comment]
Thank you @chaudharydeepak 
Seems like It could be addressed in stream. I'll post here once a new snapshot is available.

**Comment by olegz** [comment]
@chaudharydeepak try it now, the latest snapshots of spring-cloud-stream are available - https://repo.spring.io/libs-snapshot-local/org/springframework/cloud/spring-cloud-stream/3.1.3-SNAPSHOT/
LMK

**Comment by chaudharydeepak** [comment]
@olegz - I tried, same outcome. Attached herein is log.
[app_2.log](https://github.com/spring-projects-experimental/spring-native/files/6245040/app_2.log)

![Screen Shot 2021-04-01 at 11 37 57 AM](https://user-images.githubusercontent.com/931030/113326193-f6721a00-92de-11eb-8736-ef9584192c19.png)

**Comment by chaudharydeepak** [comment]
Then, if I provided the required proxy configuration to satisfy above error via
```
-H:DynamicProxyConfigurationFiles=classes/proxy-config.json
```
as:
```
[
  [
    "org.springframework.cloud.stream.annotation.EnableBinding",
    "org.springframework.core.annotation.SynthesizedAnnotation"
  ]
]
```

, then in run logs I see following[<i>from 3.1.3 snapshot of SCS</i>], which is cool!
```
...
2021-04-01 12:44:29.661  WARN 8626 --- [           main] o.s.c.s.config.BindingBeansRegistrar     : Failed to proxy EnableBinding annotation. If you are using functional programming style, ignore this warning, otherwise, annotation-based programming model is not and will not be suppported in native images.
...
```

however, now, the run fails with this:
```
2021-04-01 12:44:29.665 ERROR 8626 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'IntegrationConfigurationBeanFactoryPostProcessor': Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.integration.config.IntegrationConfigurationBeanFactoryPostProcessor]: No default constructor found; nested exception is java.lang.NoSuchMethodException: org.springframework.integration.config.IntegrationConfigurationBeanFactoryPostProcessor.<init>()
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateBean(AbstractAutowireCapableBeanFactory.java:1316) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1214) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:564) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:524) ~[na:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[na:na]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[na:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[na:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:213) ~[na:na]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:135) ~[na:na]
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:746) ~[na:na]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:564) ~[na:na]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144) ~[na:na]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:769) ~[com.dccorp.nativeapps.nativecloudstreamsapplication:2.4.4]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:761) ~[com.dccorp.nativeapps.nativecloudstreamsapplication:2.4.4]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:426) ~[com.dccorp.nativeapps.nativecloudstreamsapplication:2.4.4]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:326) ~[com.dccorp.nativeapps.nativecloudstreamsapplication:2.4.4]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1313) ~[com.dccorp.nativeapps.nativecloudstreamsapplication:2.4.4]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1302) ~[com.dccorp.nativeapps.nativecloudstreamsapplication:2.4.4]
	at com.dccorp.nativeapps.NativeCloudStreamsApplication.main(NativeCloudStreamsApplication.java:10) ~[com.dccorp.nativeapps.nativecloudstreamsapplication:0.0.1-SNAPSHOT]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.integration.config.IntegrationConfigurationBeanFactoryPostProcessor]: No default constructor found; nested exception is java.lang.NoSuchMethodException: org.springframework.integration.config.IntegrationConfigurationBeanFactoryPostProcessor.<init>()
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:83) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateBean(AbstractAutowireCapableBeanFactory.java:1308) ~[na:na]
	... 18 common frames omitted
Caused by: java.lang.NoSuchMethodException: org.springframework.integration.config.IntegrationConfigurationBeanFactoryPostProcessor.<init>()
	at java.lang.Class.getConstructor0(DynamicHub.java:3349) ~[na:na]
	at java.lang.Class.getDeclaredConstructor(DynamicHub.java:2553) ~[na:na]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:78) ~[na:na]
	... 19 common frames omitted
```

attached herein are logs:
[app.log](https://github.com/spring-projects-experimental/spring-native/files/6245456/app.log)

**Comment by olegz** [comment]
@chaudharydeepak so I am changing the `catch (Exception e) {` to `catch (Throwable e) {` (didn't realize that `com.oracle.svm.core.jdk.UnsupportedFeatureError` was actually an Error. Anyway, the SNAPSHOTS are in process (will let you know). That should let you avoid doing this `-H:DynamicProxyConfigurationFiles=classes/proxy-config.json` i believe.


As for the other error, can you please check what version of `spring-integration` you have as it seems to be like a version missmatch.

**Comment by chaudharydeepak** [comment]
@olegz - Thank you. Spring Integration version is auto pulled in as 5.4.5:

![Screen Shot 2021-04-01 at 1 22 04 PM](https://user-images.githubusercontent.com/931030/113337342-4657dd80-92ed-11eb-9cc3-14e97539e5d8.png)

**Comment by artembilan** [comment]
Spring Integration is undergoing some internal changes at the moment to avoid reflection for configuration as much as possible: https://github.com/spring-projects/spring-integration/issues/3502. 
The fix will be available in the upcoming Spring Integration `5.5`.
We also probably will add some hints into this project for more coverage. 

Will link respective Spring Integration issue for Native back to here when it it is there...

**Comment by olegz** [comment]
Ok, I was just chatting with @artembilan (the lead for spring-integration) and unfortunately there are still some issues that need to be worked out with regard to native support for spring-integration as shown [here](https://github.com/spring-projects/spring-integration/issues/3502). Perhaps @sdeleuze can chime in and see if there is anything that can be done for the spring-integration issue?

Also, new snapshots should be available, so refresh.

**Comment by chaudharydeepak** [comment]
Thank you @olegz @artembilan . I will hold-on then up until 5.5 is released. I tried bringing latest milestone, just to see if that helps.

Also I refreshed SCS 3.1.3-Snapshot - and I don't have to provide any Hints / proxy configs etc anymore :).

Thank you.

**Comment by artembilan** [comment]
For better traceability: https://github.com/spring-projects-experimental/spring-native/issues/469

**Comment by magnus-larsson** [comment]
I have a Spring Cloud Stream application that I tries to native compile with Spring Native 0.9.2 following the advice in this issue. I have upgraded to Spring Integration 5.5.0-RC1 and Spring Cloud-Stream 3.1.3-SNAPSHOT which resolves most of the problems. I'm now stuck on use of the `StreamBridge`. When starting up the native compiled application I get:

```
 2021-04-20 10:33:00.990  WARN 1 --- [           main] onfigReactiveWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'productCompositeServiceApplication': Unsatisfied dependency expressed through field 'integration'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'productCompositeIntegration' defined in class path resource [se/magnus/microservices/composite/product/services/ProductCompositeIntegration.class]: Unsatisfied dependency expressed through constructor parameter 3; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'streamBridgeUtils': Unsatisfied dependency expressed through field 'bindingService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'bindingService' defined in class path resource [org/springframework/cloud/stream/config/BindingServiceConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.cloud.stream.binding.BindingService]: Factory method 'bindingService' threw exception; nested exception is javax.validation.ValidationException: HV000183: Unable to initialize 'javax.el.ExpressionFactory'. Check that you have the EL dependencies on the classpath, or use ParameterMessageInterpolator instead
``` 

The constructor parameter #3 mentioned in the error message above is the `StreamBridge`. Any suggestions?

**Comment by olegz** [comment]
@magnus-larsson what version of stream are you using at the moment. The reason why I am asking is that there were quite a few fixes recently to accommodate native, so i'd suggest grabbing the latest snapshot.

It would also help if you can have a small bare minimum all pushed somewhere in github so we can take a look

**Comment by artembilan** [comment]
NOTE: It is better to use Spring Integration `5.5.0-SNAPSHOT` as well. Plus the latest SNAPSHOT for Spring Native `0.10.0`.

> Unable to initialize 'javax.el.ExpressionFactory'

Doesn't look like your problem is relevant somehow to Spring Cloud Stream or Spring Integration.
Not sure what is that though, but just want to narrow the boundaries for what to investigate and possible to fix...

**Comment by sdeleuze** [comment]
Should be fixed by b6877b0152d00941b7a25b3e19bd210e8d763de6 which provide a sample. If more is needed, please comment here with a repro project attached after checking you are using the same versions than the new `cloud-stream-kafka` sample.
